document/document.go

```diff
@@ -853,6 +853,52 @@ func (d Document) AddHyperlink(url string) common.Hyperlink {
 	return d.docRels.AddHyperlink(url)
 }
 
+func bookmarks(bc *wml.EG_ContentBlockContent) []Bookmark {
+	ret := []Bookmark{}
+
+	// bookmarks within paragraphs
+	for _, p := range bc.P {
+		for _, ec := range p.EG_PContent {
+			for _, ecr := range ec.EG_ContentRunContent {
+				for _, re := range ecr.EG_RunLevelElts {
+					for _, rm := range re.EG_RangeMarkupElements {
+						if rm.BookmarkStart != nil {
+							ret = append(ret, Bookmark{rm.BookmarkStart})
+						}
+					}
+				}
+			}
+		}
+	}
+	// bookmarks within block runs
+	for _, re := range bc.EG_RunLevelElts {
+		for _, rm := range re.EG_RangeMarkupElements {
+			if rm.BookmarkStart != nil {
+				ret = append(ret, Bookmark{rm.BookmarkStart})
+			}
+		}
+	}
+	// bookmarks within tables, potentially nested
+	for _, tbl := range bc.Tbl {
+		for _, crc := range tbl.EG_ContentRowContent {
+			for _, tr := range crc.Tr {
+				for _, ccc := range tr.EG_ContentCellContent {
+					for _, tc := range ccc.Tc {
+						for _, ble := range tc.EG_BlockLevelElts {
+							for _, bc := range ble.EG_ContentBlockContent {
+								for _, b := range bookmarks(bc) {
```

Comment:
Is there any guard for infinite recursion?  Like if bookmark refers to another bookmark that then refers to itself ?
Would make sense to have a cache of visited nodes to avoid that.